File: topos/graphs/cpg/builder.py
"""
CPG Builder
-----------

Construct a CodePropertyGraph from a UAST root by composing the AST, CFG,
DDG, and CDG edge families into a single labeled multigraph
(Yamaguchi et al., arxiv:1909.03496).

Algorithm
=========
1. AST family: every parent → child link from the UAST.
2. CFG family: every edge of the ControlFlowGraph, projected from
   block-level (BB -> BB) onto statement-level (last UAST stmt of source
   block → first UAST stmt of target block).
3. DDG family: each ``DependenceEdge(kind=DATA)`` from the academic PDG.
4. CDG family: each ``DependenceEdge(kind=CONTROL)`` from the academic PDG.
"""

from __future__ import annotations

from topos.graphs.cfg.object import ControlFlowGraph
from topos.graphs.cpg.models import CPGEdge, CPGEdgeKind, CPGNode
from topos.graphs.pdg.object import (
    DependenceKind,
    ProgramDependenceGraph,
)
from topos.graphs.uast.models import UASTNode


def build_cpg(
    uast_root: UASTNode,
    cfg: ControlFlowGraph | None = None,
    pdg: ProgramDependenceGraph | None = None,
    source: str = "",
) -> tuple[dict[str, CPGNode], list[CPGEdge]]:
    """Return ``(nodes, edges)`` for the CPG.

    Reuses pre-built CFG / PDG when supplied; otherwise builds them.
    ``source`` (when given) is forwarded to a freshly-built PDG so its
    data-dependence pass can recover real identifier text instead of
    falling back to per-occurrence node ids (see
    ``topos.graphs.pdg.object._identifier_name``).
    """
    if cfg is None:
        cfg = ControlFlowGraph.from_uast(uast_root)
    if pdg is None:
        pdg = ProgramDependenceGraph.from_uast(uast_root, source=source)

    nodes: dict[str, CPGNode] = {}
    _collect_nodes(uast_root, nodes)

    edges: list[CPGEdge] = []
    edges.extend(_ast_edges(uast_root))
    edges.extend(_cfg_edges(cfg))
    edges.extend(_dependence_edges(pdg))

    return nodes, edges


def _collect_nodes(root: UASTNode, out: dict[str, CPGNode]) -> None:
    stack = [root]
    while stack:
        node = stack.pop()
        key = node.id or f"anon::{id(node):x}"
        if key in out:
            continue
        out[key] = CPGNode(uast=node)
        stack.extend(node.children)


def _ast_edges(root: UASTNode) -> list[CPGEdge]:
    edges: list[CPGEdge] = []
    stack = [root]
    while stack:
        parent = stack.pop()
        parent_id = parent.id or f"anon::{id(parent):x}"
        for child in parent.children:
            child_id = child.id or f"anon::{id(child):x}"
            edges.append(
                CPGEdge(source=parent_id, target=child_id, kind=CPGEdgeKind.AST)
            )
            stack.append(child)
    return edges


def _cfg_edges(cfg: ControlFlowGraph) -> list[CPGEdge]:
    """Project block-level CFG edges down to UAST-statement-level."""
    edges: list[CPGEdge] = []
    for edge in cfg.edges:
        src_block = cfg.blocks.get(edge.source)
        dst_block = cfg.blocks.get(edge.target)
        if src_block is None or dst_block is None:
            continue
        src_stmt = src_block.statements[-1] if src_block.statements else None
        dst_stmt = dst_block.statements[0] if dst_block.statements else None
        if src_stmt is None or dst_stmt is None:
            continue
        edges.append(
            CPGEdge(
                source=src_stmt.id or "<anon>",
                target=dst_stmt.id or "<anon>",
                kind=CPGEdgeKind.CFG,
                label=str(edge.kind),
            )
        )
    return edges


def _dependence_edges(pdg: ProgramDependenceGraph) -> list[CPGEdge]:
    edges: list[CPGEdge] = []
    for dep in pdg.edges:
        kind = CPGEdgeKind.DDG if dep.kind is DependenceKind.DATA else CPGEdgeKind.CDG
        edges.append(
            CPGEdge(
                source=dep.source,
                target=dep.target,
                kind=kind,
                label=dep.var,
            )
        )
    return edges

File: crates/topos-core/src/graphs/cpg/builder.rs
//! CPG builder — construct a `CodePropertyGraph` from a UAST root by
//! composing the AST, CFG, DDG, and CDG edge families into a single
//! labeled multigraph (Yamaguchi et al., arxiv:1909.03496).
//!
//! # Algorithm
//! 1. AST family: every parent → child link from the UAST.
//! 2. CFG family: every edge of the [`ControlFlowGraph`], projected from
//!    block-level (BB → BB) onto statement-level (last UAST stmt of
//!    source block → first UAST stmt of target block).
//! 3. DDG family: each `DependenceEdge(kind=Data)` from the academic PDG.
//! 4. CDG family: each `DependenceEdge(kind=Control)` from the academic PDG.

use std::collections::HashMap;

use super::models::{CPGEdge, CPGEdgeKind, CPGNode};
use crate::graphs::cfg::object::ControlFlowGraph;
use crate::graphs::pdg::object::{DependenceKind, ProgramDependenceGraph};
use crate::graphs::uast::models::UASTNode;

/// A node key stable across a single `build_cpg` call: the UAST node's
/// own id when present, or a pointer-identity fallback otherwise.
///
/// Mirrors Python's `node.id or f"anon::{id(node):x}"`, where Python's
/// `id()` builtin is object identity — Rust's closest equivalent is the
/// node's address. In practice every node the mapper engine produces has
/// a real id; the fallback only matters for hand-built nodes (e.g.
/// `graphs::cfg::builder`'s synthetic module-callable wrapper) that
/// never appear as CFG/PDG statements themselves, so it's effectively
/// unreachable in real data, same as in Python.
fn node_key(node: &UASTNode) -> String {
    if node.id.is_empty() {
        format!("anon::{:x}", std::ptr::from_ref(node) as usize)
    } else {
        node.id.clone()
    }
}

/// Return `(nodes, edges)` for the CPG, building the CFG and PDG afresh.
pub fn build_cpg(uast_root: &UASTNode, source: &str) -> (HashMap<String, CPGNode>, Vec<CPGEdge>) {
    let cfg = ControlFlowGraph::from_uast(uast_root);
    let pdg = ProgramDependenceGraph::from_uast(uast_root, source);

    let mut nodes = HashMap::new();
    collect_nodes(uast_root, &mut nodes);

    let mut edges = Vec::new();
    edges.extend(ast_edges(uast_root));
    edges.extend(cfg_edges(&cfg));
    edges.extend(dependence_edges(&pdg));

    (nodes, edges)
}

fn collect_nodes(root: &UASTNode, out: &mut HashMap<String, CPGNode>) {
    let mut stack: Vec<&UASTNode> = vec![root];
    while let Some(node) = stack.pop() {
        let key = node_key(node);
        if out.contains_key(&key) {
            continue;
        }
        out.insert(key, CPGNode { uast: node.clone() });
        stack.extend(node.children.iter());
    }
}

fn ast_edges(root: &UASTNode) -> Vec<CPGEdge> {
    let mut edges = Vec::new();
    let mut stack: Vec<&UASTNode> = vec![root];
    while let Some(parent) = stack.pop() {
        let parent_id = node_key(parent);
        for child in &parent.children {
            let child_id = node_key(child);
            edges.push(CPGEdge::new(
                parent_id.clone(),
                child_id,
                CPGEdgeKind::Ast,
                "",
            ));
            stack.push(child);
        }
    }
    edges
}

/// Project block-level CFG edges down to UAST-statement-level.
fn cfg_edges(cfg: &ControlFlowGraph) -> Vec<CPGEdge> {
    let mut edges = Vec::new();
    for edge in &cfg.edges {
        let Some(src_block) = cfg.blocks.get(&edge.source) else {
            continue;
        };
        let Some(dst_block) = cfg.blocks.get(&edge.target) else {
            continue;
        };
        let Some(src_stmt) = src_block.statements.last() else {
            continue;
        };
        let Some(dst_stmt) = dst_block.statements.first() else {
            continue;
        };
        let source = if src_stmt.id.is_empty() {
            "<anon>".to_string()
        } else {
            src_stmt.id.clone()
        };
        let target = if dst_stmt.id.is_empty() {
            "<anon>".to_string()
        } else {
            dst_stmt.id.clone()
        };
        edges.push(CPGEdge::new(
            source,
            target,
            CPGEdgeKind::Cfg,
            edge.kind.label(),
        ));
    }
    edges
}

fn dependence_edges(pdg: &ProgramDependenceGraph) -> Vec<CPGEdge> {
    pdg.edges
        .iter()
        .map(|dep| {
            let kind = if dep.kind == DependenceKind::Data {
                CPGEdgeKind::Ddg
            } else {
                CPGEdgeKind::Cdg
            };
            CPGEdge::new(
                dep.source.clone(),
                dep.target.clone(),
                kind,
                dep.var.clone(),
            )
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graphs::ast::dispatch::parse_source;

    #[test]
    fn build_cpg_includes_all_four_edge_families_when_present() {
        let source = "def f(x):\n    if x:\n        y = 1\n    return y\n";
        let result = parse_source(source, "python", None).unwrap();
        let (nodes, edges) = build_cpg(&result.uast_root, source);

        assert!(!nodes.is_empty());
        assert!(edges.iter().any(|e| e.kind == CPGEdgeKind::Ast));
        assert!(edges.iter().any(|e| e.kind == CPGEdgeKind::Cfg));
        assert!(edges.iter().any(|e| e.kind == CPGEdgeKind::Cdg));
    }
}
